babeld/src/ubus.c

```diff
@@ -0,0 +1,373 @@
+#include <stdint.h>
+#include <sys/select.h>
+
+#include <libubox/blob.h>
+#include <libubox/blobmsg.h>
+#include <libubox/list.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+
+#include <stdlib.h>
+
+#include "babeld.h"
+#include "configuration.h"
+#include "interface.h"
+#include "kernel.h"
+#include "local.h"
+#include "message.h"
+#include "neighbour.h"
+#include "net.h"
+#include "resend.h"
+#include "route.h"
+#include "rule.h"
+#include "source.h"
+#include "util.h"
+#include "version.h"
+#include "xroute.h"
+
+#include "ubus.h"
+
+// config option for ubus bindings
+int ubus_bindings = 0;
+
+// struct definition for sorting xroutes
+struct xroute_list_entry {
+  struct list_head list;
+  struct xroute *xroute;
+};
+
+// struct definition for sorting routes
+struct route_list_entry {
+  struct list_head list;
+  struct route_stream *route;
+};
+
+// struct definition for sorting neighbours
+struct neighbour_list_entry {
+  struct list_head list;
+  struct neighbour *neighbour;
+};
+
+// pointer to ubus context
+static struct ubus_context *ctx;
+
+static void babeld_ubus_babeld_info(struct ubus_context *ctx_local,
+                                    struct ubus_object *obj,
+                                    struct ubus_request_data *req,
+                                    const char *method, struct blob_attr *msg) {
+  struct blob_buf b = {0};
+  void *prefix;
+  char host[64];
+  int ret;
+
+  blob_buf_init(&b, 0);
+  blobmsg_add_string(&b, "babeld-version", BABELD_VERSION);
+  blobmsg_add_string(&b, "my-id", format_eui64(myid));
+  if (!gethostname(host, sizeof(host)))
+    blobmsg_add_string(&b, "host", host);
+
+  ret = ubus_send_reply(ctx, req, b.head);
```

Comment:
I'm confused: do you intend to use ctx or ctx_local here? Maybe they're just the same? Because your initial code used the local one, and since the rename it uses the local one.

Maybe it makes more sense to call the local variables ctx and the global one shared_ctx or persistent_ctx?

Reply:
Yes. This was an mistake. But actually it makes no difference as you already explained. :)